### ISSUE TITLE
C++ AOT API to support graph partitioning and multi device

### ISSUE THREAD

**Issue by ShuaiShao93** [iree-org/iree#13932]
### Request description

We are interested in doing multi-gpu inference with IREE. Because we don't have nvlink for now, we can't partition the channels as collective ops. Instead we are more interested in partitioning the graph and run each subgraph on a device.

Right now, I heard there are some similar features in PJRT, but we don't use JAX and we want AOT C++ API instead. Ideally we can annotate the gpu ids for each op either on original TF model or with a dedicated API. (Or partition it automatically if the algorithm is good enough, but manual version is good enough to us). And then a simple `iree-run-mlir` can run it.

It's ok to use multiple runtime instances in the first version, but at the end, one single program should handle multiple devices.

### What component(s) does this issue relate to?

Compiler, Runtime

**Comment by stellaraccident** [comment]
What manner of partitioning are you looking for if not planning to use collectives?

**Comment by okkwon** [comment]
This sounds like a single vmfb file that can run multiple devices. I can think of two approaches.
1. Dispatch assignment to a device. This is what you described since it is a subgraph to device mapping. It may be only effective to exploit the pipeline parallelism.
2. Another way of utilizing devices is to exploit the parallelism inside a dispatch. This is more like data parallelism. The codegen process may be called "device distribution" and should be done before the work group distribution. But this still has the same problem as the SPMD partitioning; how to partition and insert data transfer (aka communication) among devices.

**Comment by ShuaiShao93** [comment]
Yeah, basically the approach 1 `pipeline parallelism` is what I want

**Comment by powderluv** [comment]
@sogartar fyi

**Comment by powderluv** [comment]
I think we have been doing something similar from python (https://github.com/nod-ai/SHARK/blob/main/shark/examples/shark_inference/sharded_bloom_large_models.py) but are moving to doing it in C++ so we can introduce cluster / node / device information.

**Comment by allieculp** [comment]
@powderluv Should we assign this to someone on your team? Looking to move this issue forward - let us know.

**Comment by allieculp** [comment]
@powderluv setting to P2 for now - let us know if this has an owner or needs a higher priority.

**Comment by parrotsky** [comment]
> Yeah, basically the approach 1 `pipeline parallelism` is what I want

I think I already implement the graph partitioning and code generation if you only need pipeline. https://github.com/parrotsky/AutoDiCE_examples
